sdk/dotnet/Pulumi.Tests/Deployment/DeploymentRunnerTests.cs

```diff
@@ -0,0 +1,174 @@
+                // repro.
+                ((Deployment.Runner)runner).WhileRunningAsync();
+            }
+        }
+
+        [Fact]
+        public async Task TerminatesEarlyOnException()
+        {
+            var deployResult = await Deployment.TryTestAsync<TerminatesEarlyOnExceptionStack>(new EmptyMocks());
+            Assert.NotNull(deployResult.Exception);
+            Assert.IsType<RunException>(deployResult.Exception!);
+            Assert.Contains("Deliberate test error", deployResult.Exception!.Message);
+            var stack = (TerminatesEarlyOnExceptionStack)deployResult.Resources[0];
+            Assert.False(stack.SlowOutput.GetValueAsync().IsCompleted);
+        }
+
+        class TerminatesEarlyOnExceptionStack : Stack
+        {
+            [Output]
+            public Output<int> SlowOutput { get; private set; }
+
+            public TerminatesEarlyOnExceptionStack()
+            {
+                Output.Create(Task.FromException<int>(new Exception("Deliberate test error")));
+                SlowOutput = Output.Create(Task.Delay(1000).ContinueWith(_ => 1));
+            }
+        }
+
+        [Fact]
+        public async Task LogsTaskDescriptions()
+        {
+            var resources = await Deployment.TestAsync<LogsTaskDescriptionsStack>(new EmptyMocks());
+            var stack = (LogsTaskDescriptionsStack)resources[0];
+            var messages = await stack.Logs;
+            for (var i = 0; i < 2; i++)
+            {
+                Assert.Contains($"Debug 0 Registering task: task{i}", messages);
+                Assert.Contains($"Debug 0 Completed task: task{i}", messages);
+            }
+        }
+
+        class LogsTaskDescriptionsStack : Stack
+        {
+            public Task<IEnumerable<string>> Logs { get; private set; }
+
+            public LogsTaskDescriptionsStack()
+            {
+                var deployment = Pulumi.Deployment.Instance.Internal;
+                var logger = new InMemoryLogger();
+                var runner = new Deployment.Runner(deployment, logger);
+
+                for (var i = 0; i < 2; i++)
+                {
+                    runner.RegisterTask($"task{i}", Task.Delay(100 + i));
+                }
+
+                this.Logs = runner.WhileRunningAsync().ContinueWith(_ => logger.Messages);
+            }
+        }
+
+        class InMemoryLogger : ILogger
+        {
+            private readonly object _lockObject = new object();
+            private readonly List<string> _messages = new List<string>();
+
+            public void Log<TState>(
+                LogLevel level,
+                EventId eventId,
+                TState state,
+                Exception exc,
+                Func<TState, Exception, string> formatter)
+            {
+                var msg = formatter(state, exc);
+                Write($"{level} {eventId} {msg}");
+            }
+
+            public IEnumerable<String> Messages {
+                get {
```

Comment:
I see you are persistent about this :)

Reply:
Go habits die hard. I write more GO here than anything else.. Damn it.